gadget/update.go

```diff
@@ -665,6 +667,143 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
 	}, nil
 }
 
+// unable to proceed with the gadget asset update, but not fatal to the refresh
+// operation itself
+var errSkipUpdateProceedRefresh = errors.New("could not identify disk for gadget asset update")
+
+// buildNewVolumeToDeviceMapping builds a DiskVolumeDeviceTraits for only the
+// system-boot role containing volume when we cannot load an existing traits
+// object from disk-mapping.json. It is meant to be used only with all UC16 /
+// UC18 installs as well as UC20 installs from before we started writing
+// disk-mapping.json during install mode.
+func buildNewVolumeToDeviceMapping(old GadgetData, laidOutVols map[string]*LaidOutVolume, preUC20 bool) (map[string]DiskVolumeDeviceTraits, error) {
+	var systemBootVolume string
+
+	// we need to pick the volume, since updates for this setup are best
+	// effort and mainly focused on the volume with system-* roles on it, we
+	// need to pick the volume with that role
+volumeLoop:
+	for volName, vol := range old.Info.Volumes {
+		for _, structure := range vol.Structure {
+			if structure.Role == SystemBoot {
+				// this is the volume
+				systemBootVolume = volName
+				break volumeLoop
+			}
+		}
+	}
+	if systemBootVolume == "" {
+		// didn't find system-boot anywhere somehow
+		if preUC20 {
+			logger.Noticef("WARNING: could not identify disk for gadget asset update of volume %s: unable to find any volume with system-boot role on it", systemBootVolume)
+			return nil, errSkipUpdateProceedRefresh
+		}
+		// shouldn't be possible on UC20
+		return nil, fmt.Errorf("unable to find any volume with system-boot, gadget is broken")
```

Comment:
usual comment to try to start errors with "cannot" as much as possible

Reply:
yes, I don't know why my brain refuses to remember this every time I go to write an error message

Reply:
fixed